File: Webfuel.App/ClientApp/src/app/api/project-support.api.ts
import { EventEmitter, Injectable, inject } from '@angular/core';
import { Observable, tap } from 'rxjs';
import { ApiService, ApiOptions } from '../core/api.service';
import { ActivatedRouteSnapshot, ResolveFn, RouterStateSnapshot } from '@angular/router';
import { IDataSource } from 'shared/common/data-source';
import { CreateProjectSupport, ProjectSupport, UpdateProjectSupport, ResendProjectSupportNotification, UpdateProjectSupportCompletion, CompleteProjectSupport, UncompleteProjectSupport, QueryProjectSupport, QueryResult } from './api.types';

@Injectable()
export class ProjectSupportApi implements IDataSource<ProjectSupport, QueryProjectSupport, CreateProjectSupport, UpdateProjectSupport> {
    constructor(private apiService: ApiService) { }
    
    public create (body: CreateProjectSupport, options?: ApiOptions): Observable<ProjectSupport> {
        return this.apiService.request<CreateProjectSupport, ProjectSupport>("POST", "api/project-support", body, options).pipe(tap(_ => this.changed.emit()));
    }
    
    public update (body: UpdateProjectSupport, options?: ApiOptions): Observable<ProjectSupport> {
        return this.apiService.request<UpdateProjectSupport, ProjectSupport>("PUT", "api/project-support", body, options).pipe(tap(_ => this.changed.emit()));
    }
    
    public resendNotification (body: ResendProjectSupportNotification, options?: ApiOptions): Observable<ProjectSupport> {
        return this.apiService.request<ResendProjectSupportNotification, ProjectSupport>("PUT", "api/project-support/resend-notification", body, options);
    }
    
    public updateCompletion (body: UpdateProjectSupportCompletion, options?: ApiOptions): Observable<ProjectSupport> {
        return this.apiService.request<UpdateProjectSupportCompletion, ProjectSupport>("PUT", "api/project-support/completion", body, options);
    }
    
    public complete (body: CompleteProjectSupport, options?: ApiOptions): Observable<ProjectSupport> {
        return this.apiService.request<CompleteProjectSupport, ProjectSupport>("PUT", "api/project-support/complete", body, options);
    }
    
    public uncomplete (body: UncompleteProjectSupport, options?: ApiOptions): Observable<ProjectSupport> {
        return this.apiService.request<UncompleteProjectSupport, ProjectSupport>("PUT", "api/project-support/uncomplete", body, options);
    }
    
    public delete (params: { id: string }, options?: ApiOptions): Observable<any> {
        return this.apiService.request<undefined, any>("DELETE", "api/project-support/" + params.id + "", undefined, options).pipe(tap(_ => this.changed.emit()));
    }
    
    public query (body: QueryProjectSupport, options?: ApiOptions): Observable<QueryResult<ProjectSupport>> {
        return this.apiService.request<QueryProjectSupport, QueryResult<ProjectSupport>>("POST", "api/project-support/query", body, options);
    }
    
    public get (params: { id: string }, options?: ApiOptions): Observable<ProjectSupport> {
        return this.apiService.request<undefined, ProjectSupport>("GET", "api/project-support/" + params.id + "", undefined, options);
    }
    
    changed = new EventEmitter<any>();
    
    // Resolvers
    
    static projectSupportResolver(param: string): ResolveFn<ProjectSupport> {
        return (route: ActivatedRouteSnapshot, state: RouterStateSnapshot): Observable<ProjectSupport> => {
            return inject(ProjectSupportApi).get({id: route.paramMap.get(param)! });
        };
    }
}

File: Webfuel.Api/Api/Project/ProjectSupportApi.cs
using MediatR;
using Microsoft.AspNetCore.Mvc;
using System.Text.Json;
using Webfuel.Domain;

namespace Webfuel.App
{
    public class UploadProjectSupportFileDTO
    {
        public required Guid ProjectId { get; set; }
    }

    [ApiService]
    [ApiDataSource]
    public static class ProjectSupportApi
    {
        public static void RegisterEndpoints(IEndpointRouteBuilder app)
        {
            // Commands

            app.MapPost("api/project-support", Create)
                .RequireIdentity();

            app.MapPut("api/project-support", Update)
                .RequireIdentity();

            app.MapPut("api/project-support/resend-notification", ResendNotification)
                .RequireIdentity();

            app.MapPut("api/project-support/completion", UpdateCompletion)
                .RequireIdentity();

            app.MapPut("api/project-support/complete", Complete)
                 .RequireIdentity();

            app.MapPut("api/project-support/uncomplete", Uncomplete)
                     .RequireIdentity();

            app.MapDelete("api/project-support/{id:guid}", Delete)
                .RequireIdentity();

            // Querys

            app.MapPost("api/project-support/query", Query)
                .RequireIdentity();

            app.MapGet("api/project-support/{id:guid}", Get)
                .RequireIdentity();

            // Files

            app.MapPost("api/project-support/select-file", SelectFile)
                 .RequireIdentity();

            app.MapPost("api/project-support/upload-file", UploadFile)
                .RequireIdentity();
        }

        public static Task<ProjectSupport> Create([FromBody] CreateProjectSupport command, IMediator mediator)
        {
            return mediator.Send(command);
        }

        public static Task<ProjectSupport> Update([FromBody] UpdateProjectSupport command, IMediator mediator)
        {
            return mediator.Send(command);
        }

        public static Task<ProjectSupport> ResendNotification([FromBody] ResendProjectSupportNotification command, IMediator mediator)
        {
            return mediator.Send(command);
        }

        public static Task<ProjectSupport> UpdateCompletion([FromBody] UpdateProjectSupportCompletion command, IMediator mediator)
        {
            return mediator.Send(command);
        }

        public static Task<ProjectSupport> Complete([FromBody] CompleteProjectSupport command, IMediator mediator)
        {
            return mediator.Send(command);
        }

        public static Task<ProjectSupport> Uncomplete([FromBody] UncompleteProjectSupport command, IMediator mediator)
        {
            return mediator.Send(command);
        }

        public static Task Delete(Guid id, IMediator mediator)
        {
            return mediator.Send(new DeleteProjectSupport { Id = id });
        }

        public static Task<QueryResult<ProjectSupport>> Query([FromBody] QueryProjectSupport command, IMediator mediator)
        {
            return mediator.Send(command);
        }

        public static async Task<ProjectSupport?> Get(Guid id, IMediator mediator)
        {
            return await mediator.Send(new GetProjectSupport { Id = id });
        }

        public static async Task<List<ProjectSupportFile>> SelectFile([FromBody] SelectProjectSupportFile command, IMediator mediator)
        {
            return await mediator.Send(command);
        }

        public static async Task<List<ProjectSupportFile>> UploadFile(HttpRequest request, IMediator mediator)
        {
            var data = request.Form["data"].ToString();
            if (String.IsNullOrEmpty(data))
                throw new InvalidOperationException("No data part found on upload project support file request");

            var command = JsonSerializer.Deserialize<UploadProjectSupportFileDTO>(data, new JsonSerializerOptions { PropertyNameCaseInsensitive = true });
            if (command == null)
                throw new InvalidOperationException("Unable to deserialize upload project support file command");

            var result = new List<ProjectSupportFile>();
            foreach (var file in request.Form.Files)
                result.Add(await mediator.Send(new UploadProjectSupportFile { ProjectId = command.ProjectId, FormFile = file }));

            return result;
        }
    }
}
